README.md
# Action Statistics

This library provides methods for storing actions with their respective times and returning the average time for each
action. The methods permit concurrent access.

## Usage

Sample usage of this library:

```python
from action_statistics import ActionStatistics

action_stats = ActionStatistics()
action_stats.addAction('''{"action":"jump", "time":100}''')  # Adds an action with time
action_stats.getStats()  # Get actions and their mean time
```

## Requirements

This library was written with Python 3.7.5.

This library uses `threading.Thread` to manage all access to the internal data store, thereby allowing for concurrent
operations. While it is possible to directly access the internal data storage object, this is strongly discouraged.
Depending on which python implementation is used, it cannot be assumed that the Global Interpreter Lock (GIL) will
safeguard access to the storage object. However, as this is python threading performance will be limited by the GIL.

## Future Considerations

There is no input error checking for input values or exception handling, this is left to the user at this time.

An initial attempt was made to use a `multiprocessing.Manager` proxy object which would bypass the Python's Global
Interpreter Lock (GIL), but preliminary sanity checks revealed a number of entries not getting stored. To see this early
stage code please see git commit `61b44847`.

# Install

This assumes you have Python 3.7.5 and `pip` installed.

```shell
pip install <path or URL of this repo>
```

# Test

Run the `test.py` script.

```shell
python3 test.py
```

The test execution time will vary some due to random sleep intervals introduced to test for concurrency issues.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Visioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## 0.1.0 - 2020-12-03

### Added
- Initial version.



test.py
"""Basic tests for this module."""
import threading
import unittest
from json import loads, dumps
from random import randint
from time import sleep

from action_statistics import ActionStatistics


class TestInit(unittest.TestCase):
    """Test the ActionStatistics.__init__ method."""

    def test_init(self):
        """Test the most basic init of the ActionStatistics class."""
        ActionStatistics()


class TestConcurrent(unittest.TestCase):
    """Test concurrency."""

    @staticmethod
    def sub_process(action_stats: ActionStatistics, add_action: dict, sleep_window=0):
        """Action to be taken by a subprocess.

        Parameter
        ---------
        action_stats : ActionStatistics
            Class to utilize.
        add_action : dict
            Action to add to the shared dict.
        sleep_window : int
            How long to sleep before taking any action. This is a positive integer value.
        """
        if sleep_window > 0:
            sleep(randint(1, sleep_window))

        for action in add_action:
            action_stats.addAction(dumps(action))

    def test_empty(self):
        """Test accessing the shared dictionary."""
        action_stats = ActionStatistics()
        assert action_stats.getStats() == '[]'

    def test_add_action(self):
        """Test the internal _add_action method."""
        action_stats = ActionStatistics()
        action_stats._add_action('''{"action":"jump", "time":100}''')
        action_stats._add_action('''{"action":"jump", "time":200}''')
        assert action_stats._dict == {'jump': {'total': 300, 'count': 2}}

    def test_get_stats(self):
        """Test the internal _get_stats method."""
        action_stats = ActionStatistics()
        action_stats._dict = {'jump': {'total': 100, 'count': 1}}
        assert loads(action_stats._get_stats()) == [{"action": "jump", "avg": 100}]

    def test_addAction(self):
        """Test adding an item."""
        action_stats = ActionStatistics()
        action_stats.addAction('''{"action":"jump", "time":100}''')
        sleep(2)
        assert action_stats._dict == {'jump': {'total': 100, 'count': 1}}

    def test_getStats(self):
        """Test getting an item."""
        action_stats = ActionStatistics()
        action_stats._dict = {'jump': {'total': 100, 'count': 1}}
        assert loads(action_stats.getStats()) == [{"action": "jump", "avg": 100.0}]

    def test_addAction_multiple(self):
        """Test that the average is being calculated correctly."""
        action_stats = ActionStatistics()
        action_stats.addAction('''{"action":"jump", "time":100}''')
        action_stats.addAction('''{"action":"jump", "time":200}''')
        sleep(2)
        assert action_stats.getStats() == '''[{"action": "jump", "avg": 150.0}]'''

    def test_thread(self):
        """Test a thread accessing the shared dict."""
        action_stats = ActionStatistics()
        thread = threading.Thread(target=self.sub_process, args=(
            action_stats,
            [
                {"action": "jump", "time": 100},
                {"action": "run", "time": 75},
                {"action": "jump", "time": 200}
            ],
            0
        ))
        thread.start()
        thread.join()
        sleep(2)
        assert action_stats._dict == {'jump': {'total': 100 + 200, 'count': 2}, 'run': {'total': 75, 'count': 1}}
        assert action_stats.getStats() == dumps([{"action": "jump", "avg": 150.0}, {"action": "run", "avg": 75.0}])

    def test_multiple_threads(self):
        """Test multiple threads accessing the method."""

        def trial_run():
            _action_stats = ActionStatistics()
            threads = list()
            for d in [[{'action': 'jump', 'time': x}] for x in range(100)]:
                t = threading.Thread(target=self.sub_process, args=(_action_stats, d, 5))
                threads.append(t)

            for t in threads:
                t.start()
            for t in threads:
                t.join()

            sleep(10)
            assert _action_stats._dict['jump']['total'] / _action_stats._dict['jump']['count'] == 49.5
            return _action_stats

        # If the values are not in a random order, re-run the test
        for _ in range(5):
            action_stats = trial_run()
            if action_stats._dict['jump'] != [x for x in range(100)]:
                break
        assert action_stats._dict['jump'] != [x for x in range(100)]


if __name__ == '__main__':
    unittest.main()


action_statistics/ActionStatistics.py
"""ActionStatistics stores the times for each of the provided actions and returns
the average for each of the actions.
"""

import threading
from json import loads, dumps  # minor performance improvement by not calling a . sub-method
from queue import SimpleQueue
from statistics import mean
from time import sleep


class ActionStatistics:
    def __init__(self):
        self.queue_add_action = SimpleQueue()
        self.queue_get_stats = SimpleQueue()
        self.queue_get_stats_response = SimpleQueue()
        self._dict = dict()  # Should only be accessed by the thread running self.worker
        threading.Thread(target=self.worker, daemon=True).start()

    def worker(self):
        """Manages queue needs."""
        while True:
            if self.queue_add_action.empty() and self.queue_get_stats.empty():
                sleep(0.5)
                continue
            if not self.queue_add_action.empty():
                item = self.queue_add_action.get_nowait()
                if item:
                    self._add_action(item)
            if not self.queue_get_stats.empty() and self.queue_get_stats.get_nowait():
                self.queue_get_stats_response.put(self._get_stats())

    def addAction(self, action: str):
        """Add an action's time to the average for that action.

        Parameters
        ----------
        action : str
            A JSON serialized string with a format like: {"action":"jump", "time":100}. The "action" field is
            case sensitive.
        """
        self.queue_add_action.put(action)

    def getStats(self):
        """Return the average times for each action.

        Returns
        -------
        str
            JSON serialized string containing a list of the actions and their respective average times.

        Raises
        ------
        queue.Empty
            If this is unable to get a response for the query within the timeout, then a
            queue.Empty exception will be raised.
        """
        self.queue_get_stats.put(True)
        return self.queue_get_stats_response.get(block=True, timeout=5)

    def _add_action(self, action: str):
        """Add an action to the internal storage object.

        Parameter
        ---------
        action : str
            JSON serialized action to add.
        """
        action = loads(action)
        action_time = action['time']
        action = action['action']
        internal_action = self._dict.get(action, {'total': 0, 'count': 0})
        internal_action['total'] += action_time
        internal_action['count'] += 1
        self._dict[action] = internal_action

    def _get_stats(self):
        """Return the stats."""
        return dumps(
            [
                {'action': k, 'avg': v['total']/v['count']} for k, v in
                self._dict.items()
            ]
        )


action_statistics/action_statistics.py
"""ActionStatistics stores the times for each of the provided actions and returns
the average for each of the actions.
"""

import threading
from json import loads, dumps  # minor performance improvement by not calling a . sub-method
from queue import SimpleQueue
from time import sleep


class ActionStatistics:
    def __init__(self):
        self.queue_add_action = SimpleQueue()
        self.queue_get_stats = SimpleQueue()
        self.queue_get_stats_response = SimpleQueue()
        self._dict = dict()  # Should only be accessed by the thread running self.worker
        threading.Thread(target=self.worker, daemon=True).start()

    def worker(self):
        """Manages queue needs."""
        while True:
            if self.queue_add_action.empty() and self.queue_get_stats.empty():
                sleep(0.5)
                continue
            if not self.queue_add_action.empty():
                item = self.queue_add_action.get_nowait()
                if item:
                    self._add_action(item)
            if not self.queue_get_stats.empty() and self.queue_get_stats.get_nowait():
                self.queue_get_stats_response.put(self._get_stats())

    def addAction(self, action: str):
        """Add an action's time to the average for that action.

        Parameters
        ----------
        action : str
            A JSON serialized string with a format like: {"action":"jump", "time":100}. The "action" field is
            case sensitive.
        """
        self.queue_add_action.put(action)

    def getStats(self):
        """Return the average times for each action.

        Returns
        -------
        str
            JSON serialized string containing a list of the actions and their respective average times with the format:
            [{"action":"jump", "avg":150}, {"action":"run", "avg":75}]

        Raises
        ------
        queue.Empty
            If this is unable to get a response for the query within the timeout, then a
            queue.Empty exception will be raised.
        """
        self.queue_get_stats.put(True)
        return self.queue_get_stats_response.get(block=True, timeout=5)

    def _add_action(self, action: str):
        """Add an action to the internal storage object.

        Parameter
        ---------
        action : str
            JSON serialized action to add.
        """
        action = loads(action)
        action_time = action['time']
        action = action['action']
        internal_action = self._dict.get(action, {'total': 0, 'count': 0})
        internal_action['total'] += action_time
        internal_action['count'] += 1
        self._dict[action] = internal_action

    def _get_stats(self):
        """Return the stats."""
        return dumps(
            [
                {'action': k, 'avg': v['total'] / v['count']} for k, v in
                self._dict.items()
            ]
        )
